docs/proposals/secret-annotation-new-app.md

```diff
@@ -0,0 +1,60 @@
+Design proposal to implement [Use build source secret based on annotation on
+Secret](https://trello.com/c/NoVpS2OS/1059-5-use-build-source-secret-based-on-annotation-on-secret-builds)
+
+(The card goal is for oc new-app/new-build to discover a relevant pre-existing
+Secret and attempt to use it when a source repo requires authentication.  The
+intention is to do this without adding yet more CLI arguments to
+new-app/new-build.)
+
+
+Design proposal:
+----------------
+
+- A Secret may include an annotation with a well-known name.  The name
+  *io.openshift.match-urls* is proposed.
+
+- A valid annotation value consists of zero or more comma-separated
+  "optionally-wildcarded URL patterns" (defined below).
+
+- If a source repository requires authentication, and a Secret exists which is
+  useful to new-app (e.g. it is a basic auth secret), and any of its
+  "optionally-wildcarded URL patterns" matches the source repository URL, oc
+  new-app/new-build will use that Secret to authenticate against the repository.
+  If the new-app/new-build process succeeds and a BuildConfig is created, the BC
+  sourceSecret will be set to refer to the same Secret.
+
+- The optionally-wildcarded URL patterns are proposed to be very similar to
+  Chrome's [Match
+  Patterns](https://developer.chrome.com/extensions/match_patterns).  I believe
+  these provide the appropriate level of flexibility, are as intuitive as these
+  things can be, cover relevant security considerations in their design, and can
+  be implemented reasonably straightforwardly.  Proposed differences from the
+  referenced spec would include:
+  - the '<all_urls>' pattern would not be implemented.
+  - permitted schemes would be 'http', 'https' and 'ssh'.
```

Comment:
the secret can also be used to provide a gitconfig for cloning, it's possible you have gitconfig customization you want to do for a particular repo that does not specify a credential that's being passed over the wire.